bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java

```diff
@@ -15,252 +15,296 @@
+     * 4002 and it processes any notifications received. The task runs continuously in a loop until the
+     * thread is externally interrupted.
+     *
+     * <li>In case of status notifications it forwards the message to the Thing handler listener.</li>
+     * <li>In case of discovery notifications it forwards the message to the discovery listener.</li>
+     * <li>If there is neither a Thing handler listener, nor a discovery listener, it logs an error.</li>
+     */
+    private void serverThreadTask() {
+        synchronized (serverLock) {
+            try {
+                logger.trace("Server thread started.");
+                ByteBuffer buffer = ByteBuffer.allocate(1024);
 
-                try (MulticastSocket sendSocket = new MulticastSocket(socketAddress)) {
-                    sendSocket.setSoTimeout(INTERFACE_TIMEOUT_SEC * 1000);
-                    sendSocket.setReuseAddress(true);
-                    sendSocket.setBroadcast(true);
-                    sendSocket.setTimeToLive(2);
-                    sendSocket.joinGroup(new InetSocketAddress(broadcastAddress, RESPONSE_PORT), intf);
+                while (!serverStopFlag) {
+                    try (DatagramChannel channel = DatagramChannel.open()
+                            .setOption(StandardSocketOptions.SO_REUSEADDR, true)
+                            .bind(new InetSocketAddress(RESPONSE_PORT))) {
 
-                    byte[] requestData = DISCOVER_REQUEST.getBytes();
+                        while (!serverStopFlag) {
+                            String sourceIp = "";
+                            try {
+                                SocketAddress socketAddress = channel.receive(buffer.clear());
+                                if ((socketAddress instanceof InetSocketAddress inetSocketAddress)
+                                        && (inetSocketAddress.getAddress() instanceof InetAddress inetAddress)) {
+                                    sourceIp = inetAddress.getHostAddress();
+                                } else {
+                                    logger.debug("Receive() - bad socketAddress={}", socketAddress);
+                                    return;
+                                }
+                            } catch (ClosedByInterruptException e) {
+                                // thrown if 'Thread.interrupt()' is called during 'channel.receive()'
+                                logger.trace(
+                                        "Receive exception=ClosedByInterruptException, isInterrupted={}, serverStopFlag={}",
+                                        Thread.currentThread().isInterrupted(), serverStopFlag);
+                                Thread.interrupted(); // clear 'interrupted' flag
+                                if (serverStopFlag) {
+                                    return;
+                                } else {
+                                    break;
+                                }
```

Comment:
Double check: if the thread is interrupted, the serverStopFlag is not (yet) set to true, so the break is called. 
Calling `break` means exiting the while loop, not returning to the caller!

Is it intentional that the rest of the method is executed when the thread is interrupted?

The rest of the method seems 0 lines, so maybe there is no difference, but a return is more clear?

Reply:
There are two nested while loops. The outer loop creates a Datagram channel, and the inner loop listens for network input on the channel. If there is a transient network stack issue that interrupts the network IO then we exit the inner loop, recreate the channel, and continue listening. Whereas if OH / Karaf is unloading the component then the server stop flag will exit from both inner and outer loops and terminate the server thread entirely.

Reply:
Aha. But I guess you are saying that we could just break the inner loop and the outer loop would continue anyway. I will make a change.